modules/library/src/main/scala/zio/elasticsearch/query/ElasticIntervalQuery.scala

```diff
@@ -0,0 +1,356 @@
+          analyzer.map("analyzer" -> Str(_)),
+          useField.map("use_field" -> _.toString.toJson)
+        ).flatten: _*
+      )
+    )
+}
+private[elasticsearch] final case class IntervalRegexp[S](
+  pattern: Regexp[S],
+  analyzer: Option[String],
+  useField: Option[Field[_, _]]
+) extends IntervalRule
+    with HasAnalyzer[IntervalRegexp[S]]
+    with HasUseField[IntervalRegexp[S]] {
+
+  override def analyzer(value: String): IntervalRegexp[S]      = copy(analyzer = Some(value))
+  override def useField(field: Field[_, _]): IntervalRegexp[S] = copy(useField = Some(field))
+
+  private[elasticsearch] def toJson: Json =
+    Obj(
+      "regexp" -> Obj(
+        Chunk(
+          Some("pattern" -> pattern.toJson(None)),
+          analyzer.map("analyzer" -> _.toJson),
+          useField.map("use_field" -> _.toString.toJson)
+        ).flatten: _*
+      )
+    )
+}
+
+private[elasticsearch] final case class IntervalWildcard[S](
+  pattern: String,
+  analyzer: Option[String],
+  useField: Option[Field[_, _]]
+) extends IntervalRule
+    with HasAnalyzer[IntervalWildcard[S]]
+    with HasUseField[IntervalWildcard[S]] {
+
+  override def analyzer(value: String): IntervalWildcard[S]      = copy(analyzer = Some(value))
+  override def useField(field: Field[_, _]): IntervalWildcard[S] = copy(useField = Some(field))
+
+  private[elasticsearch] def toJson: Json =
+    Obj(
+      "wildcard" -> Obj(
+        Chunk(
+          Some("pattern" -> pattern.toJson),
+          analyzer.map("analyzer" -> _.toJson),
+          useField.map("use_field" -> _.toString.toJson)
+        ).flatten: _*
+      )
+    )
+}
+
+object ElasticIntervalQuery {
+
+  def intervalAllOf[S](intervals: NonEmptyChunk[IntervalRule]): IntervalAllOf[S] =
+    IntervalAllOf(intervals, maxGaps = None, ordered = None, filter = None)
+
+  def intervalAnyOf[S](intervals: NonEmptyChunk[IntervalRule]): IntervalAnyOf[S] =
+    IntervalAnyOf(intervals, filter = None)
+
+  def intervalContains[S](pattern: String): IntervalWildcard[S] =
+    IntervalWildcard(s"*$pattern*", analyzer = None, useField = None)
+
+  def intervalEndsWith[S](pattern: String): IntervalWildcard[S] =
+    IntervalWildcard(s"*$pattern", analyzer = None, useField = None)
+
+  def intervalFilter[S](
+    after: Option[IntervalRule] = None,
+    before: Option[IntervalRule] = None,
+    containedBy: Option[IntervalRule] = None,
+    containing: Option[IntervalRule] = None,
+    notContainedBy: Option[IntervalRule] = None,
+    notContaining: Option[IntervalRule] = None,
+    notOverlapping: Option[IntervalRule] = None,
+    overlapping: Option[IntervalRule] = None,
+    script: Option[Json] = None
+  ): Option[IntervalFilter[S]] = {
+
+    val filter: IntervalFilter[S] = IntervalFilter(
+      after,
+      before,
+      containedBy,
+      containing,
+      notContainedBy,
+      notContaining,
+      notOverlapping,
+      overlapping,
+      script
+    )
+
+    val isEmpty = Seq(
```

Comment:
Let's use `List(...).forAll(_.isEmpty)` and inline it somehow. No need for `isEmpty` val I would say.